docs/config.md
| 配置项                    | 描述                                                                                                                                                                    | 默认值               |
|:-----------------------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------------|
| open_driver            | 开启浏览器运行，若更新无数据可开启此模式，较消耗性能                                                                                                                                            | False             |
| open_epg               | 开启EPG功能，支持频道显示预告内容                                                                                                                                                    | True              |
| open_empty_category    | 开启无结果频道分类，自动归类至底部                                                                                                                                                     | False             |
| open_filter_resolution | 开启分辨率过滤，低于最小分辨率（min_resolution）的接口将会被过滤，GUI用户需要手动安装FFmpeg，程序会自动调用FFmpeg获取接口分辨率，推荐开启，虽然会增加测速阶段耗时，但能更有效地区分是否可播放的接口                                                      | True              |
| open_filter_speed      | 开启速率过滤，低于最小速率（min_speed）的接口将会被过滤                                                                                                                                      | True              |
| open_hotel             | 开启酒店源功能，关闭后所有酒店源工作模式都将关闭                                                                                                                                              | False             |
| open_hotel_foodie      | 开启 Foodie 酒店源工作模式                                                                                                                                                     | True              |
| open_hotel_fofa        | 开启 FOFA、ZoomEye 酒店源工作模式                                                                                                                                               | False             |
| open_local             | 开启本地源功能，将使用模板文件与本地源文件中的数据                                                                                                                                             | True              |
| open_m3u_result        | 开启转换生成 m3u 文件类型结果链接，支持显示频道图标                                                                                                                                          | True              |
| open_multicast         | 开启组播源功能，关闭后所有组播源工作模式都将关闭                                                                                                                                              | False             |
| open_multicast_foodie  | 开启 Foodie 组播源工作模式                                                                                                                                                     | True              |
| open_multicast_fofa    | 开启 FOFA 组播源工作模式                                                                                                                                                       | False             |
| open_online_search     | 开启关键字搜索源功能                                                                                                                                                            | False             |
| open_request           | 开启查询请求，数据来源于网络（仅针对酒店源与组播源）                                                                                                                                            | False             |
| open_rtmp              | 开启RTMP推流功能，需要安装FFmpeg，利用本地带宽提升接口播放体验                                                                                                                                  | False             |
| open_service           | 开启页面服务，用于控制是否启动结果页面服务；如果使用青龙等平台部署，有专门设定的定时任务，需要更新完成后停止运行，可以关闭该功能                                                                                                      | True              |
| open_speed_test        | 开启测速功能，获取响应时间、速率、分辨率                                                                                                                                                  | True              |
| open_subscribe         | 开启订阅源功能                                                                                                                                                               | False             |
| open_supply            | 开启补偿机制模式，用于控制当频道接口数量不足时，自动将不满足条件（例如低于最小速率）但可能可用的接口添加至结果中，从而避免结果为空的情况                                                                                                  | True              |
| open_update            | 开启更新，用于控制是否更新接口，若关闭则所有工作模式（获取接口和测速）均停止                                                                                                                                | True              |
| open_update_time       | 开启显示更新时间                                                                                                                                                              | True              |
| open_url_info          | 开启显示接口说明信息，用于控制是否显示接口来源、分辨率、协议类型等信息，为$符号后的内容，播放软件使用该信息对接口进行描述，若部分播放器（如PotPlayer）不支持解析导致无法播放可关闭                                                                        | False             |
| open_use_cache         | 开启使用本地缓存数据，适用于查询请求失败场景（仅针对酒店源与组播源）                                                                                                                                    | True              |
| open_history           | 开启使用历史更新结果（包含模板与结果文件的接口），合并至本次更新中                                                                                                                                     | True              |
| open_headers           | 开启使用M3U内含的请求头验证信息，用于测速等操作，注意：只有个别播放器支持播放这类含验证信息的接口，默认为关闭                                                                                                              | False             |
| app_port               | 页面服务端口，用于控制页面服务的端口号                                                                                                                                                   | 8000              |
| cdn_url                | CDN代理加速地址，用于订阅源、频道图标等资源的加速访问                                                                                                                                          |                   |
| final_file             | 生成结果文件路径                                                                                                                                                              | output/result.txt |
| hotel_num              | 结果中偏好的酒店源接口数量                                                                                                                                                         | 10                |
| hotel_page_num         | 酒店地区获取分页数量                                                                                                                                                            | 1                 |
| hotel_region_list      | 酒店源地区列表，"全部"表示所有地区                                                                                                                                                    | 全部                |
| isp                    | 接口运营商，用于控制结果中只包含填写的运营商类型，支持关键字过滤，英文逗号分隔，不填写表示不指定运营商                                                                                                                   |                   |
| ipv4_num               | 结果中偏好的 IPv4 接口数量                                                                                                                                                      | 5                 |
| ipv6_num               | 结果中偏好的 IPv6 接口数量                                                                                                                                                      | 5                 |
| ipv6_support           | 强制认为当前网络支持IPv6，跳过检测                                                                                                                                                   | False             |
| ipv_type               | 生成结果中接口的协议类型，可选值：ipv4、ipv6、全部、all                                                                                                                                     | 全部                |
| ipv_type_prefer        | 接口协议类型偏好，优先将该类型的接口排在结果前面，可选值：ipv4、ipv6、自动、auto                                                                                                                        | ipv6,ipv4         |
| location               | 接口归属地，用于控制结果只包含填写的归属地类型，支持关键字过滤，英文逗号分隔，不填写表示不指定归属地，建议使用靠近使用者的归属地，能提升播放体验                                                                                              |                   |
| local_file             | 本地源文件路径                                                                                                                                                               | config/local.txt  |
| local_num              | 结果中偏好的本地源接口数量                                                                                                                                                         | 10                |
| min_resolution         | 接口最小分辨率，需要开启 open_filter_resolution 才能生效                                                                                                                              | 1920x1080         |
| max_resolution         | 接口最大分辨率，需要开启 open_filter_resolution 才能生效                                                                                                                              | 1920x1080         |
| min_speed              | 接口最小速率（单位M/s），需要开启 open_filter_speed 才能生效                                                                                                                             | 0.5               |
| multicast_num          | 结果中偏好的组播源接口数量                                                                                                                                                         | 10                |
| multicast_page_num     | 组播地区获取分页数量                                                                                                                                                            | 1                 |
| multicast_region_list  | 组播源地区列表，"全部"表示所有地区                                                                                                                                                    | 全部                |
| online_search_num      | 结果中偏好的关键字搜索接口数量                                                                                                                                                       | 0                 |
| online_search_page_num | 关键字搜索频道获取分页数量                                                                                                                                                         | 1                 |
| origin_type_prefer     | 结果偏好的接口来源，结果优先按该顺序进行排序，逗号分隔，例如：local,hotel,multicast,subscribe,online_search；local：本地源，hotel：酒店源，multicast：组播源，subscribe：订阅源，online_search：关键字搜索；不填写则表示不指定来源，按照接口速率排序 |                   |
| recent_days            | 获取最近时间范围内更新的接口（单位天），适当减小可避免出现匹配问题                                                                                                                                     | 30                |
| request_timeout        | 查询请求超时时长，单位秒(s)，用于控制查询接口文本链接的超时时长以及重试时长，调整此值能优化更新时间                                                                                                                   | 10                |
| speed_test_limit       | 同时执行测速的接口数量，用于控制测速阶段的并发数量，数值越大测速所需时间越短，负载较高，结果可能不准确；数值越小测速所需时间越长，低负载，结果较准确；调整此值能优化更新时间                                                                                | 10                |
| speed_test_timeout     | 单个接口测速超时时长，单位秒(s)；数值越大测速所需时间越长，能提高获取接口数量，但质量会有所下降；数值越小测速所需时间越短，能获取低延时的接口，质量较好；调整此值能优化更新时间                                                                             | 10                |
| speed_test_filter_host | 测速阶段使用Host地址进行过滤，相同Host地址的频道将共用测速数据，开启后可大幅减少测速所需时间，但可能会导致测速结果不准确                                                                                                      | False             |
| source_file            | 模板文件路径                                                                                                                                                                | config/demo.txt   |
| subscribe_num          | 结果中偏好的订阅源接口数量                                                                                                                                                         | 10                |
| time_zone              | 时区，可用于控制更新时间显示的时区，可选值：Asia/Shanghai 或其它时区编码                                                                                                                           | Asia/Shanghai     |
| urls_limit             | 单个频道接口数量                                                                                                                                                              | 10                |
| update_interval        | 定时执行更新时间间隔，单位小时，设置0或空则只运行一次，不作用于工作流                                                                                                                                   | 12                |
| update_time_position   | 更新时间显示位置，需要开启 open_update_time 才能生效，可选值：top、bottom，top: 显示于结果顶部，bottom: 显示于结果底部                                                                                       | top               |

CHANGELOG.md
# 更新日志（Changelog）

## v1.7.2

### 2025/5/26

### 🚀 新功能 ###

---

- 新增支持设置`定时更新间隔`，`命令行` `GUI` `Docker`均可实现定时间隔更新，可通过配置`update_interval`设置执行更新任务时间的间隔，默认
  `12小时`，不作用于工作流，工作流依旧每日
  `6点与18点`执行更新

### 🌟 优化 ###

---

- 更新频道别名数据，欢迎提供更多别名，参见：💖 [频道别名收集计划](https://github.com/Guovin/iptv-api/discussions/1082)

### 🐛 修复 ###

---

- 修复公网推流`APP_HOST`配置应用（#1094）
- 修复部分场景下未开启测速获取结果未保存问题（#1092）
- 修复频道缓存结果`解冻`失败
- 修复部分设备无法打开`GUI`界面

### 🗑️ 移除 ###

---

- 移除Docker`UPDATE_CRON`环境变量，请使用`config/config.ini`文件中`update_interval`参数控制更新时间间隔

<details>
  <summary>English</summary>

### 🚀 New Features ###

---

- Added support for setting `scheduled update interval`. Both `CLI`, `GUI`, and `Docker` now support scheduled interval
  updates. You can set the interval for executing update tasks via the `update_interval` configuration. The default is
  `12 hours`. This does not apply to workflows, which still update daily at
  `6:00 and 18:00`.

### 🌟 Optimization ###

---

- Updated channel alias data. Contributions for more aliases are welcome. See:
  💖 [Channel Alias Collection Plan](https://github.com/Guovin/iptv-api/discussions/1082)

### 🐛 Bug Fixes ###

---

- Fixed the application of the public streaming APP_HOST configuration (#1094)
- Fixed the issue where results were not saved when speed test was not enabled in some scenarios (#1092)
- Fixed failure to "unfreeze" channel cache results
- Fixed some devices unable to open the `GUI` interface

### 🗑️ Removal ###

---

- Removed Docker `UPDATE_CRON` environment variable. Please use the `update_interval` parameter in the
  `config/config.ini` file to control the update interval.

</details>

## v1.7.1

### 2025/5/9

### 🚀 新功能 ###

---

- 新增支持获取接口`归属地`与`运营商`（利用`纯真IP数据库`实现），支持关键字过滤，可通过配置`location`与`isp`
  生成想要的结果，建议优先使用靠近使用环境的归属地与本机网络运营商，以提升播放效果（#1058）
- 新增支持无需开启测速的情况下，可对接口进行`排序`，输出结果日志

### 🌟 优化 ###

---

- 优化`IPv6`结果进入缓存
- 调整`冻结结果的阈值`，加入`最大延迟`与`最小速率`限制
- 调整默认配置`ipv_type_prefer = auto`，即根据网络环境自动选择排序IPv4与IPv6结果的优先级
- 频道结果日志文件更名为`result.log`
- 更新部分配置参数描述

### 🐛 修复 ###

---

- 修复`IPv6含参数结果`匹配问题（#1048）
- 修复白名单生成结果失败（#1055）

### 🗑️ 移除 ###

---

- 移除无效的`IPv6订阅源`

> [!NOTE]
> 有小伙伴对部署后首次更新时间变长有疑问，其实这是正常的。
> 因为从`v1.7.0`开始，为了提升频道测速准确性，默认对接口进行全量测速。
> 目前首次运行一般`30分钟`左右，如果是新增的频道比较多首次运行时间会比较长。
> 但这并不会影响使用，由于默认模板已经内置了部分更新结果（`output/cache.pkl.gz`），部署后可立即访问使用。
> 同时测速阶段可根据历史数据跳过无效接口，无需担心，后续更新所需时间会明显减少。
> 如果你介意，可开启Host共享模式（`speed_test_filter_host = True`），相同Host的接口会共享测速结果，可以大幅降低测速所需时间，但结果准确性也会下降。

<details>
  <summary>English</summary>

### 🚀 New Features ###

---

- Added support for obtaining interface `location` and `ISP` (implemented using the `IPIP database`), supports keyword
  filtering. You can configure `location` and `isp` to generate desired results. It is recommended to prioritize the
  location and ISP close to the usage environment to improve playback performance (#1058).
- Added support for sorting interfaces and outputting result logs without enabling speed testing.

### 🌟 Optimizations ###

---

- Optimized caching of `IPv6` results.
- Adjusted the `frozen result threshold` by adding `maximum latency` and `minimum speed` limits.
- Adjusted the default configuration `ipv_type_prefer = auto`, which automatically prioritizes sorting of IPv4 and IPv6
  results based on the network environment.
- Renamed the channel result log file to `result.log`.
- Updated descriptions of some configuration parameters.

### 🐛 Bug Fixes ###

---

- Fixed the issue with matching `IPv6 results with parameters` (#1048).
- Fixed the failure to generate whitelist results (#1055).

### 🗑️ Removals ###

---

- Removed invalid `IPv6 subscription sources`.

> [!NOTE]
> Some users have raised concerns about the longer initial update time after deployment. This is actually normal.
> Starting from `v1.7.0`, to improve the accuracy of channel speed tests,
> full speed testing of interfaces is enabled by default.
> The first run usually takes about `30 minutes`. If there are many new channels, the initial run time may be longer.
> However, this does not affect usage, as the default template already includes some pre-updated results
> (`output/cache.pkl.gz`), allowing immediate access after deployment.
> During the speed test phase, invalid interfaces can be skipped based on historical data, so there is no need to worry.
> Subsequent updates will take significantly less time.
> If you are concerned, you can enable Host sharing mode (`speed_test_filter_host = True`), where interfaces with the
> same Host share speed test results. This can greatly reduce the time required for speed testing,
> but the accuracy of the results may decrease.

</details>

## v1.7.0

### 2025/5/1

### 🚀 新功能 ###

---

- 新增`频道别名`功能（`config/alias.txt`），提升频道名称匹配能力
- 新增`EPG`功能（订阅文件配置`config/epg.txt`），显示频道预告信息
- 支持`回放类接口`获取与生成
- 新增`历史结果`的冻结与解冻，`冻结`：无效结果不参与测速，`解冻`：无结果时自动解冻重新测速
- 新增`最大分辨率`限制`max_resolution`
- 支持含`请求头`信息接口测速与生成，需播放器支持才可播放，可通过`open_headers`控制是否开启
- 新增测速并发数量配置`speed_test_limit`，实现控制测速负载压力
- 新增`Host数据共享`配置`speed_test_filter_host`，实现相同Host地址接口可共享测速结果
- 新增`推流统计`GUI按钮

### 🌟 优化 ###

---

- 重构`测速与排序`逻辑，适配更多类型接口的测速（#1009）
- 提供`内置结果`，解决首次运行等待期间无结果问题（可能不稳定，建议使用更新后结果）
- 优化接口测速默认为`全接口测速`，解决Host共享结果部分接口测速不准确问题
- 调整测速结果以`速率`排序，`分辨率`不再参与，解决部分低速率接口在前的问题
- 默认开启`推流`，调整`HLS`分片配置，推荐使用`HLS`接口，缓解卡顿情况
- 重构接口`额外信息`处理逻辑
- 测速相关配置项更名为`speed_test_*`，修改输出日志文案
- 调整默认最低接口速率为`0.5M/s`
- 更新黑名单，增加无效接口与`音频`接口

### 🐛 修复 ###

---

- 修复工作流运行问题，更换使用最新`ubuntu`版本（#1032）
- 修复`M3U`订阅源白名单失效问题（#1019）
- 修复部分`组播源`测速问题（#1026）
- 修复接口协议分类结果生成失败问题

### 🗑️ 移除 ###

---

- 移除部分失效订阅源
- 移除代理更新功能`open_proxy`
- 移除保留模式`open_keep_all`
- 移除重复执行`sort_duplicate_limit`

<details>
  <summary>English</summary>

### 🚀 New Features ###

---

- Added `Channel Alias` feature (`config/alias.txt`) to improve channel name matching.
- Added `EPG` feature (subscription file configuration `config/epg.txt`) to display channel program information.
- Support for `Playback Interface` retrieval and generation.

utils/config.py
import configparser
import os
import re
import shutil
import sys


def resource_path(relative_path, persistent=False):
    """
    Get the resource path
    """
    base_path = os.path.abspath(".")
    total_path = os.path.join(base_path, relative_path)
    if persistent or os.path.exists(total_path):
        return total_path
    else:
        try:
            base_path = sys._MEIPASS
            return os.path.join(base_path, relative_path)
        except Exception:
            return total_path


def get_resolution_value(resolution_str):
    """
    Get resolution value from string
    """
    pattern = r"(\d+)[xX*](\d+)"
    match = re.search(pattern, resolution_str)
    if match:
        width, height = map(int, match.groups())
        return width * height
    else:
        return 0


class ConfigManager:

    def __init__(self):
        self.config = None
        self.load()

    def __getattr__(self, name, *args, **kwargs):
        return getattr(self.config, name, *args, **kwargs)

    @property
    def open_service(self):
        return self.config.getboolean("Settings", "open_service", fallback=True)

    @property
    def open_update(self):
        return self.config.getboolean("Settings", "open_update", fallback=True)

    @property
    def open_use_cache(self):
        return self.config.getboolean("Settings", "open_use_cache", fallback=True)

    @property
    def open_request(self):
        return self.config.getboolean("Settings", "open_request", fallback=False)

    @property
    def open_filter_speed(self):
        return self.config.getboolean(
            "Settings", "open_filter_speed", fallback=True
        )

    @property
    def open_filter_resolution(self):
        return self.config.getboolean(
            "Settings", "open_filter_resolution", fallback=True
        )

    @property
    def ipv_type(self):
        return self.config.get("Settings", "ipv_type", fallback="全部").lower()

    @property
    def open_ipv6(self):
        return (
                "ipv6" in self.ipv_type or "all" in self.ipv_type or "全部" in self.ipv_type
        )

    @property
    def ipv_type_prefer(self):
        return [
            ipv_type_value.lower()
            for ipv_type in self.config.get(
                "Settings", "ipv_type_prefer", fallback=""
            ).split(",")
            if (ipv_type_value := ipv_type.strip())
        ]

    @property
    def ipv4_num(self):
        try:
            return self.config.getint("Settings", "ipv4_num", fallback=5)
        except:
            return ""

    @property
    def ipv6_num(self):
        try:
            return self.config.getint("Settings", "ipv6_num", fallback=5)
        except:
            return ""

    @property
    def ipv6_support(self):
        return self.config.getboolean("Settings", "ipv6_support", fallback=False)

    @property
    def ipv_limit(self):
        return {
            "all": self.urls_limit,
            "ipv4": self.ipv4_num,
            "ipv6": self.ipv6_num,
        }

    @property
    def origin_type_prefer(self):
        return [
            origin_value.lower()
            for origin in self.config.get(
                "Settings",
                "origin_type_prefer",
                fallback="",
            ).split(",")
            if (origin_value := origin.strip())
        ]

    @property
    def hotel_num(self):
        return self.config.getint("Settings", "hotel_num", fallback=10)

    @property
    def multicast_num(self):
        return self.config.getint("Settings", "multicast_num", fallback=10)

    @property
    def subscribe_num(self):
        return self.config.getint("Settings", "subscribe_num", fallback=10)

    @property
    def online_search_num(self):
        return self.config.getint("Settings", "online_search_num", fallback=10)

    @property
    def source_limits(self):
        return {
            "all": self.urls_limit,
            "local": self.local_num,
            "hotel": self.hotel_num,
            "multicast": self.multicast_num,
            "subscribe": self.subscribe_num,
            "online_search": self.online_search_num,
        }

    @property
    def min_speed(self):
        return self.config.getfloat("Settings", "min_speed", fallback=0.5)

    @property
    def min_resolution(self):
        return self.config.get("Settings", "min_resolution", fallback="1920x1080")

    @property
    def min_resolution_value(self):
        return get_resolution_value(self.min_resolution)

    @property
    def max_resolution(self):
        return self.config.get("Settings", "max_resolution", fallback="1920x1080")

    @property
    def max_resolution_value(self):
        return get_resolution_value(self.max_resolution)

    @property
    def urls_limit(self):
        return self.config.getint("Settings", "urls_limit", fallback=30)

    @property
    def open_url_info(self):
        return self.config.getboolean("Settings", "open_url_info", fallback=True)

    @property
    def recent_days(self):
        return self.config.getint("Settings", "recent_days", fallback=30)

    @property
    def source_file(self):
        return self.config.get("Settings", "source_file", fallback="config/demo.txt")

    @property
    def final_file(self):
        return self.config.get("Settings", "final_file", fallback="output/result.txt")

    @property
    def open_m3u_result(self):
        return self.config.getboolean("Settings", "open_m3u_result", fallback=True)

    @property
    def open_subscribe(self):
        return self.config.getboolean("Settings", f"open_subscribe", fallback=True)

    @property
    def open_hotel(self):
        return self.config.getboolean("Settings", f"open_hotel", fallback=True)

    @property
    def open_hotel_fofa(self):
        return self.config.getboolean("Settings", f"open_hotel_fofa", fallback=True)

    @property
    def open_hotel_foodie(self):
        return self.config.getboolean("Settings", f"open_hotel_foodie", fallback=True)

    @property
    def open_multicast(self):
        return self.config.getboolean("Settings", f"open_multicast", fallback=True)

    @property
    def open_multicast_fofa(self):
        return self.config.getboolean("Settings", f"open_multicast_fofa", fallback=True)

    @property
    def open_multicast_foodie(self):
        return self.config.getboolean(
            "Settings", f"open_multicast_foodie", fallback=True
        )

    @property
    def open_online_search(self):
        return self.config.getboolean("Settings", f"open_online_search", fallback=True)

    @property
    def open_method(self):
        return {
            "epg": self.open_epg,
            "local": self.open_local,
            "subscribe": self.open_subscribe,
            "hotel": self.open_hotel,
            "multicast": self.open_multicast,
            "online_search": self.open_online_search,
            "hotel_fofa": self.open_hotel and self.open_hotel_fofa,
            "hotel_foodie": self.open_hotel and self.open_hotel_foodie,
            "multicast_fofa": self.open_multicast and self.open_multicast_fofa,
            "multicast_foodie": self.open_multicast and self.open_multicast_foodie,
        }

    @property
    def open_history(self):
        return self.config.getboolean("Settings", "open_history", fallback=True)

    @property
    def open_speed_test(self):
        return self.config.getboolean("Settings", "open_speed_test", fallback=True)

    @property
    def open_update_time(self):
        return self.config.getboolean("Settings", "open_update_time", fallback=True)

    @property
    def multicast_region_list(self):
        return [
            region.strip()
            for region in self.config.get(
                "Settings", "multicast_region_list", fallback="全部"
            ).split(",")
            if region.strip()
        ]

    @property
    def hotel_region_list(self):
        return [
            region.strip()
            for region in self.config.get(
                "Settings", "hotel_region_list", fallback="全部"
            ).split(",")
            if region.strip()
        ]

    @property
    def request_timeout(self):
        return self.config.getint("Settings", "request_timeout", fallback=10)

    @property
    def speed_test_timeout(self):
        return self.config.getint("Settings", "speed_test_timeout", fallback=10)

    @property
    def open_driver(self):
        return self.config.getboolean(
            "Settings", "open_driver", fallback=False
        )

    @property
    def hotel_page_num(self):
        return self.config.getint("Settings", "hotel_page_num", fallback=1)

    @property
    def multicast_page_num(self):
        return self.config.getint("Settings", "multicast_page_num", fallback=1)

    @property
    def online_search_page_num(self):
        return config.getint("Settings", "online_search_page_num", fallback=1)

    @property
    def open_empty_category(self):
        return self.config.getboolean("Settings", "open_empty_category", fallback=True)

    @property
    def app_host(self):
        return os.getenv("APP_HOST") or self.config.get("Settings", "app_host", fallback="http://localhost")

    @property
    def app_port(self):
        return os.getenv("APP_PORT") or self.config.getint("Settings", "app_port", fallback=8000)

    @property
    def open_supply(self):
        return self.config.getboolean("Settings", "open_supply", fallback=True)

    @property
    def update_time_position(self):
        return self.config.get("Settings", "update_time_position", fallback="top")

    @property
    def time_zone(self):
        return self.config.get("Settings", "time_zone", fallback="Asia/Shanghai")

    @property
    def open_local(self):
        return self.config.getboolean("Settings", "open_local", fallback=True)

    @property
    def local_file(self):
        return self.config.get("Settings", "local_file", fallback="config/local.txt")

    @property
    def local_num(self):
        return self.config.getint("Settings", "local_num", fallback=10)

    @property
    def speed_test_filter_host(self):
        return self.config.getboolean("Settings", "speed_test_filter_host", fallback=False)

    @property
    def cdn_url(self):
        return self.config.get("Settings", "cdn_url", fallback="")

    @property
    def open_rtmp(self):
        return not os.getenv("GITHUB_ACTIONS") and self.config.getboolean("Settings", "open_rtmp", fallback=True)

    @property
    def open_headers(self):
        return self.config.getboolean("Settings", "open_headers", fallback=False)

    @property
    def open_epg(self):
        return self.config.getboolean("Settings", "open_epg", fallback=True)

    @property
    def speed_test_limit(self):
        return self.config.getint("Settings", "speed_test_limit", fallback=10)

    def load(self):
        """
        Load the config
        """
        self.config = configparser.ConfigParser()
        user_config_path = resource_path("config/user_config.ini")
        default_config_path = resource_path("config/config.ini")

        # user config overwrites default config
        config_files = [default_config_path, user_config_path]
        for config_file in config_files:
            if os.path.exists(config_file):
                with open(config_file, "r", encoding="utf-8") as f:
                    self.config.read_file(f)

    def set(self, section, key, value):
        """
        Set the config
        """
        self.config.set(section, key, value)

    def save(self):
        """
        Save config with write
        """
        user_config_file = "config/" + (
            "user_config.ini"
            if os.path.exists(resource_path("user_config.ini"))
            else "config.ini"
        )
        user_config_path = resource_path(user_config_file, persistent=True)
        if not os.path.exists(user_config_path):
            os.makedirs(os.path.dirname(user_config_path), exist_ok=True)
        with open(user_config_path, "w", encoding="utf-8") as configfile:
            self.config.write(configfile)

    def copy(self, path="config"):
        """
        Copy config files to current directory
        """
        dest_folder = os.path.join(os.getcwd(), path)
        try:
            src_dir = resource_path(path)
            if os.path.exists(src_dir):
                if not os.path.exists(dest_folder):
                    os.makedirs(dest_folder, exist_ok=True)

                for root, _, files in os.walk(src_dir):
                    for file in files:
                        src_file_path = os.path.join(root, file)
                        relative_path = os.path.relpath(src_file_path, src_dir)
                        dest_file_path = os.path.join(dest_folder, relative_path)

                        dest_file_dir = os.path.dirname(dest_file_path)
                        if not os.path.exists(dest_file_dir):
                            os.makedirs(dest_file_dir, exist_ok=True)

                        if not os.path.exists(dest_file_path):
                            shutil.copy(src_file_path, dest_file_path)
        except Exception as e:
            print(f"Failed to copy files: {str(e)}")


config = ConfigManager()
